src/inmanta/data/__init__.py

```diff
@@ -3358,6 +3358,31 @@ async def mark_all_as_non_primary(cls, connection: Optional[asyncpg.connection.C
         """
         await cls._execute_query(query, connection=connection)
 
+    @classmethod
+    async def clean_up(cls, connection: Optional[asyncpg.connection.Connection] = None) -> None:
+        query = """
+DELETE FROM public.agent AS a
+WHERE (environment, name) NOT IN (
+    SELECT DISTINCT environment_id as environment, agent as name
+    FROM (
+        -- agent is in the agent map
+        SELECT e.id as environment_id, map.key as agent
+        FROM public.environment e
+        CROSS JOIN LATERAL jsonb_each(e.settings->'autostart_agent_map') AS map(key, value)
+    ) in_agent_map
```

Comment:
This sub-query must filter on the correct environment.

Reply:
I seems I was too quick here. This query seems to perform better than my suggestion. You can ignore this one.

Reply:
Ah. I see. You are using the (environment, name) index on the agent table here.